### PR TITLE
Article describing procedure for using a Converged Network Adapter and iSCSI to install and boot Harvester

### PR DESCRIPTION
**Pull request by jeff-radick-suse**
See https://github.com/harvester/harvester/issues/5183

In limited cases, systems having a Converged Network Adapter supporting iSCSI
can make use of storage on an iSCSI server instead of locally attached storage
for the disk that Harvester installs onto and boots from.

Some modifications to the boot configuration must be made manually
for this to work.

This KB article explains the limited use case, context,
and the procedure to make use of this.

This is *not* describing the general case of using non-local storage for Harvester installations
in all situations.

### CODE REVIEW

**Review by bk201 — APPROVED** [review]
Thanks, a very useful article!

**Review by ibrokethecloud — APPROVED** [review]
lgtm. thanks.

**Comment by jillian-maroket** [comment]
@jeff-radick-suse I'll review this today.

File: kb/2024-02-19/booting_harvester_via_iscsi_ibt.md [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+2. Configure system firmware to boot via iSCSI using the available CNA
+3. Boot the Harvester install image and install to the iSCSI device
+4. On first Harvester boot after installation, edit the kernel boot parameters in the GRUB kernel command line
+5. Permanently edit the GRUB configuration file in the normally read-only partition
+
+Steps for this need to be performed interactively, so a fully automated / unattended installation is **not** possible at this time.
+
+It is **very important** to be aware that if you subsequently upgrade Harvester,
+the upgrade **will overwrite** the "permanently edited" GRUB parameters,
+so you will have to re-edit those parameters by hand.
```
Comment (jillian-maroket) [inline kb/2024-02-19/booting_harvester_via_iscsi_ibt.md:155]:
```suggestion
## Procedure

The following is a summary of the procedure. Individual steps, which are described in the following sections, must be performed interactively. A fully automated installation is **not** possible at this time.

1. Provision storage for your Harvester node on your iSCSI server system.
1. Configure system firmware to boot via iSCSI using the available CNA.
1. Boot the Harvester install image and install to the iSCSI device.
1. On first Harvester boot after installation, edit the kernel boot parameters in the GRUB kernel command line.
1. Permanently edit the GRUB configuration file in the normally read-only partition.

:::info important
The boot configuration changes will persist across node reboots but **not** across system upgrades, which will overwrite the GRUB parameters. 
:::
```

File: kb/2024-02-19/booting_harvester_via_iscsi_ibt.md [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+
+## Provision storage for your Harvester node on your iSCSI server system
```
Comment (jillian-maroket) [inline kb/2024-02-19/booting_harvester_via_iscsi_ibt.md:157]:
```suggestion
### 1. Provision storage for your Harvester node on your iSCSI server system.
```

File: kb/2024-02-19/booting_harvester_via_iscsi_ibt.md [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+
+```
+    iqn.2024-02.com.example:cluster1-node0-boot-disk
+```
+
+There are lots of variations of this format, and this is just an example.
+
+The correct name to use should be chosen in consultation with the administrator of your storage server and storage area network.
+
+## Configure system firmware to boot via iSCSI using the available CNA
```
Comment (jillian-maroket) [inline kb/2024-02-19/booting_harvester_via_iscsi_ibt.md:191]:
```suggestion
### 2. Configure system firmware to boot via iSCSI using the available CNA.
```

File: kb/2024-02-19/booting_harvester_via_iscsi_ibt.md [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+An example document link is provided in the References section.
+Other vendors should provide similar documentation.
+
+The typical things you need to configure are:
+* Enable UEFI boot
+* Configure iSCSI initiator and target parameters
+* Enable the iSCSI device in the boot menu
+* Set the boot order so that your system will boot from the iSCSI device
+
+## Boot the Harvester install image and install to the iSCSI device
```
Comment (jillian-maroket) [inline kb/2024-02-19/booting_harvester_via_iscsi_ibt.md:213]:
```suggestion
### Boot the Harvester install image and install to the iSCSI device
```

File: kb/2024-02-19/booting_harvester_via_iscsi_ibt.md [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+This can be done by whatever means you would normally use to load the Harvester install image.
+
+The Harvester installer _should_ automatically "see" the iSCSI device in the dialog where you chose the installation destination.
+Choose this device to install.
+
+Installation should proceed and complete normally.
+
+When installation completes, your system should reboot.
+
+## On first boot, edit kernel boot parameters in the GRUB kernel command line
```
Comment (jillian-maroket) [inline kb/2024-02-19/booting_harvester_via_iscsi_ibt.md:224]:
```suggestion
### 4. On first boot, edit kernel boot parameters in the GRUB kernel command line.
```

File: kb/2024-02-19/booting_harvester_via_iscsi_ibt.md [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+  linux ($loopdev)$kernel $kernelcmd ${extra_cmdline} ${extra_active_cmdline} rd.iscsi.firmware rd.iscsi.ibft
+```
+
+At this point, type Ctrl-X to resume booting with the modified kernel command line.
+
+Now the node should come up normally, and finish with the normal Harvester console screen that shows the cluster and node IP addresses and status.
+
+The the node should operate normally now **but** the kernel boot argument changes will not be preserved across a reboot unless you perform the next step.
+
+## Permanently edit the GRUB configuration file
```
Comment (jillian-maroket) [inline kb/2024-02-19/booting_harvester_via_iscsi_ibt.md:273]:
```suggestion
### 5. Permanently edit the GRUB configuration file.
```

File: kb/2024-02-19/booting_harvester_via_iscsi_ibt.md [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+
+
+# References
+
+1. [SCSI](https://en.wikipedia.org/wiki/SCSI) provides an overview of SCSI and contains references to additional material.
+2. [iSCSI](https://en.wikipedia.org/wiki/ISCSI) provides an overview of iSCSI and contains references to additional material.
+3. [Converged Network Adapter](https://en.wikipedia.org/wiki/Converged_network_adapter) provides a summary of CNAs and references to additional material.
+4. [Harvester Docuementation](https://docs.harvesterhci.io/v1.2/troubleshooting/os/#how-to-permanently-edit-kernel-parameters) provides a general description of how to permanently edit kernel parameters to be used when booting a Harvester node.
+5. [Dell PowerEdge R630 Owner's Manual](https://www.dell.com/support/manuals/en-us/poweredge-r630/r630_om_pub/uefi-iscsi-settings?guid=guid-adc7d625-5c7b-469d-ba9c-4a2c704fcc49&lang=en-us) This is an example of relevant vendor documentation.  Other vendors such as HPE, IBM, Lenovo, etc should provide comparable documentation, though the details will vary.
```
Comment (jillian-maroket) [inline kb/2024-02-19/booting_harvester_via_iscsi_ibt.md:333]:
```suggestion
## References

- [SCSI](https://en.wikipedia.org/wiki/SCSI) provides an overview of SCSI and contains references to additional material.
- [iSCSI](https://en.wikipedia.org/wiki/ISCSI) provides an overview of iSCSI and contains references to additional material.
- [Converged Network Adapter](https://en.wikipedia.org/wiki/Converged_network_adapter) provides a summary of CNAs and references to additional material.
- [Harvester Docuementation](https://docs.harvesterhci.io/v1.2/troubleshooting/os/#how-to-permanently-edit-kernel-parameters) provides a general description of how to permanently edit kernel parameters to be used when booting a Harvester node.
- [Dell PowerEdge R630 Owner's Manual](https://www.dell.com/support/manuals/en-us/poweredge-r630/r630_om_pub/uefi-iscsi-settings?guid=guid-adc7d625-5c7b-469d-ba9c-4a2c704fcc49&lang=en-us) This is an example of relevant vendor documentation.  Other vendors such as HPE, IBM, Lenovo, etc should provide comparable documentation, though the details will vary.
```

**Review by jillian-maroket — COMMENTED** [review]
@jeff-radick-suse In this review, I focused on tightening the document structure and rewording conceptual text for brevity and clarity. I did not review the sections that describe detailed steps because I am unfamiliar with the technology. My hope is that readers can breeze through the first few sections and spend more time digesting the procedural content.

File: kb/2024-02-19/booting_harvester_via_iscsi_ibt.md [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+
+Through Harvester release 1.3.0, no explicit support has been provided for installing Harvester
+to any type of storage which is not locally attached,
+nor does it contain any explicit support for booting and running from storage which is not locally attached.
+
+However, it has been observed that systems with CNAs _might_ in fact permit Harvester to do just this.
+
+After some experimentation it has been determined that in _certain cases_ it _does_ work,
+_but_ the boot loader configuration of the installed system must be modified after installation,
+according to the procedure described in the following sections of this document.
```
[the remark this answers is not in the document]
Comment (jeff-radick-suse) [reply kb/2024-02-19/booting_harvester_via_iscsi_ibt.md]: I agree with all of your suggestions and have incorporated them.  Thank you for your thoughtful reading of this document.

**Comment by jillian-maroket** [comment]
@jeff-radick-suse A few suggestions seem to have been missed. Please check the **Files changed** tab.

File: kb/2024-02-19/booting_harvester_via_iscsi_ibt.md [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+
+# Requirements and Limitations
+
+The procedures described in this document do not apply to all circumstances.
+
+To the extent possible, necessary requirements and limitations are described here.
```
[the remark this answers is not in the document]
Comment (jeff-radick-suse) [reply kb/2024-02-19/booting_harvester_via_iscsi_ibt.md]: done

**Comment by jeff-radick-suse** [comment]
Thank you for catching my error. I *think* I have them all this time, but don't hesitate to let me know if I still missed some, or if you find other things that need correction.

**Comment by jeff-radick-suse** [comment]
This fork of the harvesterhci.io repo is messed up so I am unable to merge it into the main repo.  I don't know how to recover it.  After trying various things that didn't work I'm going to cut my losses and start over . 
I'm going to close this PR, delete the fork, make a new fork, add the file to the new fork, and make a new PR.

**Pull request closed without merging**